py-rse/style.Rmd

```diff
@@ -287,755 +349,875 @@ n = count(lines)
+    or turned into logging statements (Section \@ref(py-rse-errors-logging)).
 
-In-place operators save us some typing.
-They also make the intention clearer,
-and most importantly,
-they make it harder to get complex assignments wrong.
-For example:
+-   The variables `ln`, `i`, and `j` should be given clearer names.
+
+-   The outer loop in `count` loops over the indices of the line list
+    rather than over the lines.
+    It should do the latter
+    (which will allow us to get rid of the variable `i`).
+
+-   Rather than counting how often a word occurs in the list of stop words with `stops.count`,
+    we can turn the stop words into a set and use `in` to check words.
+    This will be more readable *and* more efficient.
+
+-   There's no reason to store the result of `line.split` in a temporary variable:
+    the inner loop of `count` can use it directly.
+
+-   Since the set of stop words is a global variable,
+    it should be written in upper case.
+
+-   We should use `+=` to increment the counter `n`.
+
+-   Rather than reading the input into a list of lines and then looping over that,
+    we can give `count` a stream and have it process the lines one by one.
+
+-   Since we might want to use `count` in other programs some day,
+    we should put the two lines at the bottom that handle input into a conditional
+    so that they aren't executed when this script is imported.
+
+After making all these changes,
+our program looks like this:
 
 ```python
-samples[least_factor_index, max(current_offset, offset_limit)] *= scaling_factor
+import sys
+
+
+STOPS = {'a', 'A', 'the', 'The', 'and'}
+
+
+def count(reader):
+    n = 0
+    for line in reader:
+        for word in line.split():
+            if word in STOPS:
+                n += 1
+    return n
+
+
+if __name__ == '__main__':
+    n = count(sys.stdin)
+    print('number', n)
 ```
 
-is much easier to read than the equivalent expression:
+Reading code in order to find bugs and suggest improvements like these
+is called [code review][code-review].
+Study after study over more than 40 years has shown that code review is the most effective way to find bugs in software
+@Faga1976,@Faga1986,@Cohe2010,@Bacc2013.
+It is also a great way to transfer knowledge between programmers:
+reading someone else's code critically will give you lots of ideas about what you could do better,
+and highlight things that you should probably stop doing as well.
+
+Despite this,
+code review still isn't common in research software development.
+This is partly a chicken-and-egg problem---people don't do it
+because other people don't do it @Sega2005.
+It is also more difficult to do in specialized scientific fields:
+in order for review to be useful,
+reviewers need to understand the problem domain well enough to comment on algorithms and design choices
+rather than indentation and variable naming,
+and the number of people who can do that for a research project is often very small @Petr2014.
```

Comment:
Also, in fields where software development is used merely as a tool and not an end goal itself, it is rarely seen as important enough by supervisors, so they will likely not ask any of their grad students to spend time on it. Or they only have one grad student that can code in the first place and see not point is getting more.

Reply:
note added.